File: backend/core/resume_clean.py
"""Normalize noisy text extracted from PDF/DOCX resumes and job descriptions."""
from __future__ import annotations

import re

from core.contact_extract import EMAIL_RE, GITHUB_RE, LEETCODE_RE, LINKEDIN_RE, PHONE_RE, URL_RE

# PDF font-encoding artifacts and stray punctuation common in extracted resumes.
CID_RE = re.compile(r"\(?cid:\s*\d+\s*\)?", re.IGNORECASE)
CID_COMMA_DEBRIS_RE = re.compile(r"\s*(?:,\s*)+$", re.MULTILINE)
CONTROL_RE = re.compile(r"[\x00-\x08\x0b\x0c\x0e-\x1f\x7f-\x9f]")
ZERO_WIDTH_RE = re.compile(r"[\u200b-\u200d\ufeff]")
REPLACEMENT_CHAR_RE = re.compile(r"\ufffd")
JUNK_SYMBOLS_RE = re.compile(r"[§¶†‡•◦·▪▫●○◆◇■□]+")
LIGATURE_RE = re.compile(r"[ﬁﬂﬀﬃﬄ]")
SMART_QUOTE_RE = str.maketrans({"“": '"', "”": '"', "‘": "'", "’": "'", "–": "-", "—": "-", "−": "-"})
HYPHEN_BREAK_RE = re.compile(r"(\w)-\n(\w)")
LETTER_SPACED_RE = re.compile(r"\b(?:[A-Za-z]\s){4,}[A-Za-z]\b")
_PROTECT_PATTERNS = (
    EMAIL_RE,
    URL_RE,
    LINKEDIN_RE,
    GITHUB_RE,
    LEETCODE_RE,
    PHONE_RE,
)


def _merge_spans(spans: list[tuple[int, int, str]]) -> list[tuple[int, int, str]]:
    if not spans:
        return []
    spans.sort(key=lambda item: (item[0], -(item[1] - item[0])))
    merged: list[tuple[int, int, str]] = []
    for start, end, value in spans:
        if merged and start < merged[-1][1]:
            continue
        merged.append((start, end, value))
    return merged


def _protect_contact_spans(text: str) -> tuple[str, list[str]]:
    spans: list[tuple[int, int, str]] = []
    for pattern in _PROTECT_PATTERNS:
        for match in pattern.finditer(text):
            value = match.group(0).strip()
            if pattern is PHONE_RE and len(re.sub(r"\D", "", value)) < 10:
                continue
            spans.append((match.start(), match.end(), value))
    merged = _merge_spans(spans)
    protected: list[str] = []
    out = text
    for idx, (start, end, value) in enumerate(reversed(merged)):
        token = f"__RESUME_PROTECTED_{len(merged) - 1 - idx}__"
        protected.insert(0, value)
        out = out[:start] + token + out[end:]
    return out, protected


def _restore_contact_spans(text: str, protected: list[str]) -> str:
    out = text
    for idx, value in enumerate(protected):
        out = out.replace(f"__RESUME_PROTECTED_{idx}__", value)
    return out


def _fix_ligatures(text: str) -> str:
    return (
        text.replace("ﬁ", "fi")
        .replace("ﬂ", "fl")
        .replace("ﬀ", "ff")
        .replace("ﬃ", "ffi")
        .replace("ﬄ", "ffl")
    )


def _fix_hyphenated_line_breaks(text: str) -> str:
    return HYPHEN_BREAK_RE.sub(r"\1\2", text)


def _join_wrapped_lines(text: str) -> str:
    lines = text.split("\n")
    joined: list[str] = []
    buffer = ""
    for raw_line in lines:
        line = raw_line.strip()
        if not line:
            if buffer:
                joined.append(buffer)
                buffer = ""
            joined.append("")
            continue
        if not buffer:
            buffer = line
            continue
        if (
            buffer[-1] not in ".!?:;"
            and line[0].islower()
            and not re.match(r"^[\d•\-*]", line)
            and len(buffer) < 120
        ):
            buffer = f"{buffer} {line}"
        else:
            joined.append(buffer)
            buffer = line
    if buffer:
        joined.append(buffer)
    return "\n".join(joined)


def _dedupe_lines(text: str) -> str:
    lines = text.split("\n")
    deduped: list[str] = []
    prev_key = None
    for line in lines:
        key = line.strip().lower()
        if key and key == prev_key:
            continue
        deduped.append(line)
        prev_key = key if key else prev_key
    return "\n".join(deduped)


def _strip_noise(text: str) -> str:
    cleaned = text.translate(SMART_QUOTE_RE)
    cleaned = _fix_ligatures(cleaned)
    cleaned = CID_RE.sub("", cleaned)
    cleaned = CID_COMMA_DEBRIS_RE.sub("", cleaned)
    cleaned = re.sub(r"^(?:,\s*)+", "", cleaned, flags=re.MULTILINE)
    cleaned = re.sub(r",\s*,+", ", ", cleaned)
    cleaned = CONTROL_RE.sub("", cleaned)
    cleaned = ZERO_WIDTH_RE.sub("", cleaned)
    cleaned = REPLACEMENT_CHAR_RE.sub("", cleaned)
    cleaned = JUNK_SYMBOLS_RE.sub(" ", cleaned)
    cleaned = cleaned.replace("\u00a0", " ")
    cleaned = cleaned.replace("\r\n", "\n").replace("\r", "\n")
    cleaned = _fix_hyphenated_line_breaks(cleaned)
    cleaned = re.sub(r"[ \t]+", " ", cleaned)
    cleaned = re.sub(r" *\n *", "\n", cleaned)
    cleaned = _join_wrapped_lines(cleaned)
    cleaned = _dedupe_lines(cleaned)
    cleaned = re.sub(r"\n{3,}", "\n\n", cleaned)
    return cleaned


def _normalize_lines(text: str) -> str:
    lines: list[str] = []
    for raw_line in text.split("\n"):
        line = raw_line.strip()
        line = re.sub(r"\s*(?:,\s*)+$", "", line).strip()
        if not line:
            if lines and lines[-1] != "":
                lines.append("")
            continue
        if LETTER_SPACED_RE.fullmatch(line):
            line = line.replace(" ", "")
        # Drop lines that are only punctuation / encoding debris.
        if re.fullmatch(r"[\W_]+", line) and not re.search(r"[@:/.]", line):
            continue
        lines.append(line)
    while lines and lines[-1] == "":
        lines.pop()
    return "\n".join(lines)


def clean_resume_text(text: str) -> str:
    """Remove PDF/DOCX extraction noise while preserving contact URLs and emails."""
    if not text:
        return ""
    protected_text, protected_values = _protect_contact_spans(text)
    cleaned = _strip_noise(protected_text)
    cleaned = _restore_contact_spans(cleaned, protected_values)
    return _normalize_lines(cleaned)


clean_document_text = clean_resume_text


def resume_preview_excerpt(text: str, *, limit: int = 500) -> str:
    cleaned = clean_resume_text(text)
    if len(cleaned) <= limit:
        return cleaned
    return cleaned[:limit].rstrip() + "…"

File: frontend/src/utils/resumeClean.js
const CID_RE = /\(?cid:\s*\d+\s*\)?/gi;
const CONTROL_RE = /[\x00-\x08\x0b\x0c\x0e-\x1f\x7f-\x9f]/g;
const ZERO_WIDTH_RE = /[\u200b-\u200d\ufeff]/g;
const REPLACEMENT_CHAR_RE = /\ufffd/g;
const JUNK_SYMBOLS_RE = /[§¶†‡•◦·▪▫●○◆◇■□]+/g;

const EMAIL_RE = /[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/g;
const URL_RE = /https?:\/\/[^\s<>"']+|(?:www\.)[^\s<>"']+/gi;
const LINKEDIN_RE = /https?:\/\/(?:[\w.-]+\.)?linkedin\.com\/in\/[\w%-]+\/?/gi;
const GITHUB_RE = /https?:\/\/(?:[\w.-]+\.)?github\.com\/[\w-]+\/?/gi;
const LEETCODE_RE = /https?:\/\/(?:[\w.-]+\.)?leetcode\.com\/(?:u\/)?[\w-]+\/?/gi;
const PHONE_RE = /(?:\+?\d{1,3}[\s.-]?)?(?:\(?\d{2,4}\)?[\s.-]?)?\d{3,4}[\s.-]?\d{3,4}(?:[\s.-]?\d{1,6})?/g;

const PROTECT_PATTERNS = [EMAIL_RE, URL_RE, LINKEDIN_RE, GITHUB_RE, LEETCODE_RE, PHONE_RE];

function mergeSpans(spans) {
  if (!spans.length) return [];
  const sorted = [...spans].sort((a, b) => a.start - b.start || b.end - b.start - (a.end - a.start));
  const merged = [];
  for (const span of sorted) {
    if (merged.length && span.start < merged[merged.length - 1].end) continue;
    merged.push(span);
  }
  return merged;
}

function protectContactSpans(text) {
  const spans = [];
  for (const pattern of PROTECT_PATTERNS) {
    pattern.lastIndex = 0;
    let match = pattern.exec(text);
    while (match) {
      const value = match[0].trim();
      if (pattern === PHONE_RE && value.replace(/\D/g, "").length < 10) {
        match = pattern.exec(text);
        continue;
      }
      spans.push({ start: match.index, end: match.index + match[0].length, value });
      match = pattern.exec(text);
    }
  }
  const merged = mergeSpans(spans);
  const protectedValues = [];
  let out = text;
  for (let i = merged.length - 1; i >= 0; i -= 1) {
    const { start, end, value } = merged[i];
    const token = `__RESUME_PROTECTED_${i}__`;
    protectedValues[i] = value;
    out = out.slice(0, start) + token + out.slice(end);
  }
  return { text: out, protectedValues };
}

function restoreContactSpans(text, protectedValues) {
  let out = text;
  protectedValues.forEach((value, idx) => {
    out = out.split(`__RESUME_PROTECTED_${idx}__`).join(value);
  });
  return out;
}

function stripNoise(text) {
  let cleaned = text;
  cleaned = cleaned.replace(CID_RE, "");
  cleaned = cleaned.replace(/\s*(?:,\s*)+$/gm, "");
  cleaned = cleaned.replace(/^(?:,\s*)+/gm, "");
  cleaned = cleaned.replace(/,\s*,+/g, ", ");
  cleaned = cleaned.replace(CONTROL_RE, "");
  cleaned = cleaned.replace(ZERO_WIDTH_RE, "");
  cleaned = cleaned.replace(REPLACEMENT_CHAR_RE, "");
  cleaned = cleaned.replace(JUNK_SYMBOLS_RE, " ");
  cleaned = cleaned.replace(/\u00a0/g, " ");
  cleaned = cleaned.replace(/\r\n/g, "\n").replace(/\r/g, "\n");
  cleaned = cleaned.replace(/[ \t]+/g, " ");
  cleaned = cleaned.replace(/ *\n */g, "\n");
  cleaned = cleaned.replace(/\n{3,}/g, "\n\n");
  return cleaned;
}

function normalizeLines(text) {
  const lines = [];
  for (const rawLine of text.split("\n")) {
    let line = rawLine.trim().replace(/\s*(?:,\s*)+$/, "").trim();
    if (!line) {
      if (lines.length && lines[lines.length - 1] !== "") lines.push("");
      continue;
    }
    if (/^[\W_]+$/.test(line) && !/[@:/.]/.test(line)) continue;
    lines.push(line);
  }
  while (lines.length && lines[lines.length - 1] === "") lines.pop();
  return lines.join("\n");
}

export function cleanResumeText(text) {
  if (!text) return "";
  const { text: protectedText, protectedValues } = protectContactSpans(text);
  let cleaned = stripNoise(protectedText);
  cleaned = restoreContactSpans(cleaned, protectedValues);
  return normalizeLines(cleaned);
}

export function resumePreviewFromUpload(data) {
  if (data?.cleaned_text) return data.cleaned_text;
  return cleanResumeText(data?.raw_text_preview || "");
}

export function resumePreviewExcerpt(text, limit = 500) {
  const cleaned = cleanResumeText(text);
  if (cleaned.length <= limit) return cleaned;
  return `${cleaned.slice(0, limit).trimEnd()}…`;
}

export function resumePreviewMeta(text) {
  const cleaned = cleanResumeText(text);
  const lines = cleaned ? cleaned.split("\n").filter(Boolean).length : 0;
  return { lines, chars: cleaned.length };
}
